src/irr/video/COpenGLPipelineCache.h

```diff
@@ -33,52 +33,75 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
 
 	void merge(uint32_t _count, const IGPUPipelineCache** _srcCaches) override
 	{
-		const std::lock_guard<std::mutex> _(m_mutex);
+		const std::lock_guard<std::mutex> _1_(m_bin_cache_mutex);
+		const std::lock_guard<std::mutex> _2_(m_parsed_cache_mutex);
 
 		for (uint32_t i = 0u; i < _count; ++i)
 		{
-			const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_cache;
-			m_cache.insert(src.begin(), src.end());
+			{
+				const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_cache;
+				m_cache.insert(src.begin(), src.end());
+			}
+			{
+				const auto& src = static_cast<const COpenGLPipelineCache*>(_srcCaches[i])->m_parsedSpirvs;
+				m_parsedSpirvs.insert(src.begin(), src.end());
+			}
 		}
 	}
 
 	core::smart_refctd_ptr<asset::ICPUPipelineCache> convertToCPUCache() const override;
 
 	COpenGLSpecializedShader::SProgramBinary find(const SCacheKey& _key, const COpenGLPipelineLayout* _layout) const
 	{
-		const std::lock_guard<std::mutex> _(m_mutex);
+		const std::lock_guard<std::mutex> _(m_bin_cache_mutex);
 
 		auto rng = m_cache.equal_range(_key);
 		for (auto it = rng.first; it != rng.second; ++it)
 			if (_layout->isCompatibleUpToSet(COpenGLPipelineLayout::DESCRIPTOR_SET_COUNT-1u, it->second.layout.get())==(COpenGLPipelineLayout::DESCRIPTOR_SET_COUNT-1u))
 				return it->second.binary;
 		return {0,nullptr};
 	}
-	//assumes that m_cache does not already contain an item with key==_key and layout fully compatible with _val.layout
-	void insert(SCacheKey&& _key, SCacheVal&& _val, const asset::ICPUBuffer* _spirv)
+	const spirv_cross::ParsedIR* findParsedSpirv(const std::array<uint64_t, 4>& _key)
 	{
-		const std::lock_guard<std::mutex> _(m_mutex);
-		/*
-		auto it = m_parsedSpirvs.find(_key.hash);
-		if (it == m_parsedSpirvs.end())
-		{
-			spirv_cross::Parser parser(reinterpret_cast<const uint32_t*>(_spirv->getPointer()), _spirv->getSize()/4ull);
-			auto parsed = parser.get_parsed_ir();
+		const std::lock_guard<std::mutex> _(m_parsed_cache_mutex);
 
-			m_parsedSpirvs.insert(it, {_key.hash,std::move(parsed)});
-		}
-		*/
+		auto found = m_parsedSpirvs.find(_key);
+		if (found!=m_parsedSpirvs.end())
+			return &found->second;
+
+		return nullptr;
+	}
+
+	//assumes that m_cache does not already contain an item with key==_key and layout fully compatible with _val.layout
+	void insert(SCacheKey&& _key, SCacheVal&& _val)
+	{
+		const std::lock_guard<std::mutex> _(m_bin_cache_mutex);
 #ifdef _IRR_DEBUG
 		assert(!find(_key, _val.layout.get()).binary);
```

Comment:
wrong assert, should be `!found || !isCompatible(found.layout.get(),_val.layout.get()`

Reply:
see code of `find()`

Reply:
`find` assumes that `layout` is compatible with the key?

Reply:
`find` doesnt find anything (returns nullptr) if passed layout is not compatible with any of entries of equal key